README.md

```diff
@@ -0,0 +1,103 @@
+  image's `init-region` to make sure these resources are in place for the
+  services that the image provides.
+  Examples include:
+  * Database schemas and credentials
+  * Credentials for messaging, external cloud services, or other facilities.
+  * Service registration, for example keystone endpoints.
+
+## Image requirements
+To enable the infrastructure described above, we require the following elements
+to be implemented for each image.
+* When deployed, each container must have access to a shared distributed
+  key-value store, like consul, etcd, or zookeeper. So far we've only enable
+  consul, but we'd like the system to allow for the possibility of others. This
+  store acts as the source of truth for all service configuration parameters.
+* Each container image implements an entrypoint called `init-region` in the
+  container image's root user PATH. This entrypoint is in charge of
+  creating/upgrading the services' external environment. Any results of this
+  setup (credentials, addresses etc) must be stored in the key-value store
+  described above.
+* Each container service has the ability to render its own in-container
+  configuration from values in the configuration key-value store. This could be
+  as simple as a wrapper script that renders a configuration file from a
+  template before starting the service process itself. In many of the
+  containers we've created, we run a `watcher` process along side the service,
+  that waits for configuration changes, re-renders configuration and restarts
+  services.  Tools like [confd](https://github.com/kelseyhightower/confd) and
+  [consul-template](https://github.com/hashicorp/consul-template) can be useful
+  for this.
+
+## Container Startup Order
+External deployment orchestration must often understand the needs of each
+service simply to determine the order in which to initialize and start each
+container.
+
+For example, say we have a container running rabbitmq, and another container
+(call it Consumer) that must have credentials on the rabbit broker to do work.
+One way to manage this is to start the rabbit container, and then run
+initialization logic for Consumer to create the user. This requires that the
+external orchestration logic perform these operations in a strict order: start
+rabbit, initialize Consumer, start Consumer. This can get pretty complicated if
+there are a lot of services providing and consuming resources like this.
+
+Axon initialization doesn't require actual interaction between services, only
+between the service and the configuration store. In the above case, instead of
+having Consumer explicitly call rabbitmq to create a user for itself, it simply
+adds configuration to the key-value store that expresses the need for that
+user. When the rabbit container eventually comes up, its own startup logic, or
+a `watcher` running alongside the rabbitmq broker service, can make sure the user
+is there.
+
```

Comment:
What is the general recommendation about services using databases? I know that currently for openstack, services require the database to be up at init-region time in order to initialize their schema. Are db's special cased? I guess in theory, a service needing schema initialization could upload a script to the config store, and the db could pick it up and execute it at a later time.